### PR TITLE
[IOTDB-5638] Support CASE WHEN expression

### PR DESCRIPTION
**Pull request by liyuheng55555**
add a new kind of expression

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    for (Expression expression : this.getExpressions()) {
+      if (!expression.isMappable(expressionTypes)) return false;
+    }
+    return true;
+  }
+
+  @Override
+  protected boolean isConstantOperandInternal() {
+    for (Expression expression : this.getExpressions()) {
+      if (!expression.isConstantOperand()) return false;
```
Comment (lancelly) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:91]: same as above.

**Review by lancelly — CHANGES_REQUESTED** [review]
PTAL~

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4 [hunk 2]
```diff
@@ -975,6 +976,23 @@ expression
...
+    | caseWhenThenExpression2
+    ;
+
+caseWhenThenExpression1
+    : CASE whenThenExpression+ (ELSE elseExpression=expression)? END
+    ;
+
+caseWhenThenExpression2
+    : CASE caseExpression=expression whenThenExpression+ (ELSE elseExpression=expression)? END
+    ;
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4:990]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    while (true) {
+      Expression expression = Expression.deserialize(byteBuffer);
+      if (expression.getExpressionType() == ExpressionType.WHEN_THEN) {
+        this.whenThenExpressions.add((WhenThenExpression) expression);
+      } else {
+        this.elseExpression = expression;
+        break;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:61]: > 

done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    result.add(elseExpression);
+    return result;
+  }
+
+  @Override
+  public <R, C> R accept(ExpressionVisitor<R, C> visitor, C context) {
+    return visitor.visitCaseWhenThenExpression(this, context);
+  }
+
+  // TODO: constructUdfExecutors
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:161]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java [hunk 5]
```diff
@@ -2430,6 +2437,58 @@ private Expression parseRoundFunction(
...
+      Expression then = parseExpression(whenThenExpressionContext.thenExpression, canUseFullPath);
+      Expression comparison = new EqualToExpression(caseExpression, when);
+      whenThenList.add(new WhenThenExpression(comparison, then));
+    }
+    Expression elseExpression = new NullOperand();
+    if (context.elseExpression != null) {
+      elseExpression = parseExpression(context.elseExpression, canUseFullPath);
+    }
+    return new CaseWhenThenExpression(whenThenList, elseExpression);
+  }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java:2483]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/transformation/dag/column/CaseWhenThenColumnTransformer.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  public CaseWhenThenColumnTransformer(
+      Type returnType,
+      List<ColumnTransformer> whenTransformers,
+      List<ColumnTransformer> thenTransformers,
+      ColumnTransformer elseTransformer) {
+    super(returnType);
+    if (whenTransformers.size() != thenTransformers.size()) {
+      throw new UnsupportedOperationException(
+          "the size of whenTransformers and thenTransformers need to be same.");
+    }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/transformation/dag/column/CaseWhenThenColumnTransformer.java:51]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionAnalyzer.java [hunk 7]
```diff
@@ -214,9 +216,40 @@ public static ResultColumn.ColumnType identifyOutputColumnType(
...
+          typeList.stream().allMatch(columnType -> columnType == ResultColumn.ColumnType.CONSTANT);
+      if (allConst) {
+        throw new SemanticException("Constant column is not supported.");
+      }
+      for (ResultColumn.ColumnType type : typeList) {
+        if (type != ResultColumn.ColumnType.CONSTANT) {
+          return type;
+        }
+      }
+      throw new IllegalArgumentException("shouldn't attach here");
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionAnalyzer.java:249]: needn't change

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java [hunk 8]
```diff
@@ -296,6 +300,52 @@ public TSDataType visitNullOperand(NullOperand nullOperand, Void context) {
...
+      }
+      TSDataType elseType = process(caseWhenThenExpression.getElseExpression(), context);
+      if (elseType != null) {
+        typeSet.add(elseType);
+      }
+      // if TEXT exist, every branch need to be TEXT
+      if (typeSet.contains(TSDataType.TEXT)) {
+        if (typeSet.stream().anyMatch(tsDataType -> tsDataType != TSDataType.TEXT)) {
+          throw new SemanticException(
+              "CASE expression: TEXT and other types cannot exist at same time");
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java:319]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java [hunk 9]
```diff
@@ -296,6 +300,52 @@ public TSDataType visitNullOperand(NullOperand nullOperand, Void context) {
...
+      // if BOOLEAN exist, every branch need to be BOOLEAN
+      if (typeSet.contains(TSDataType.BOOLEAN)) {
+        if (typeSet.stream().anyMatch(tsDataType -> tsDataType != TSDataType.BOOLEAN)) {
+          throw new SemanticException(
+              "CASE expression: BOOLEAN and other types cannot exist at same time");
+        }
+        return setExpressionType(caseWhenThenExpression, TSDataType.BOOLEAN);
+      }
+      // other 4 TSDataType can exist at same time
+      // because they can transform by Type, finally treated as DOUBLE
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java:332]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/transformation/dag/column/CaseWhenThenColumnTransformer.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    } else if (returnType instanceof FloatType) {
+      builder.writeFloat(childTransformer.getType().getFloat(column, index));
+    } else if (returnType instanceof DoubleType) {
+      builder.writeDouble(childTransformer.getType().getDouble(column, index));
+    } else if (returnType instanceof BinaryType) {
+      builder.writeBinary(childTransformer.getType().getBinary(column, index));
+    } else {
+      throw new UnsupportedOperationException("Unsupported Type");
+    }
+  }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/transformation/dag/column/CaseWhenThenColumnTransformer.java:84]: The function has been called twice in evaluate(). Removing this function may eliminate redundant code.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ExpressionAnalyzeVisitor.java [hunk 11]
```diff
@@ -32,8 +32,10 @@ public R visitExpression(Expression expression, C context) {
...
 
   List<R> getResultsFromChild(Expression expression, C context) {
-    return expression.getExpressions().stream()
-        .map(child -> process(child, context))
-        .collect(Collectors.toList());
+    List<R> result = new ArrayList<>();
+    for (Expression child : expression.getExpressions()) {
+      result.add(this.process(child, context));
+    }
+    return result;
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ExpressionAnalyzeVisitor.java:39]: restored to the version that uses stream().

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/IsDeviceViewNeedSpecialProcessVisitor.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.plan.expression.visitor;
+
+public class IsDeviceViewNeedSpecialProcessVisitor extends ReconstructVisitor<Boolean> {}
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/IsDeviceViewNeedSpecialProcessVisitor.java:22]: deleted.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ColumnTransformerVisitor.java [hunk 13]
```diff
@@ -337,6 +341,47 @@ public ColumnTransformer visitNullOperand(
...
+                TypeFactory.getType(context.getType(caseWhenThenExpression)),
+                context.originSize + context.commonTransformerList.size());
+        ColumnTransformer columnTransformer = context.hasSeen.get(caseWhenThenExpression);
+        columnTransformer.addReferenceCount();
+        context.commonTransformerList.add(columnTransformer);
+        context.leafList.add(identity);
+        context.inputDataTypes.add(context.getType(caseWhenThenExpression));
+        context.cache.put(caseWhenThenExpression, identity);
+      } else {
+        //        List<WhenThenColumnTransformer> whenThenColumnTransformers = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ColumnTransformerVisitor.java:360]: deleted.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java (hunk 1 above)
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:91]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java [hunk 14]
```diff
@@ -296,6 +300,52 @@ public TSDataType visitNullOperand(NullOperand nullOperand, Void context) {
...
+      }
+      // if TEXT exist, every branch need to be TEXT
+      if (typeSet.contains(TSDataType.TEXT)) {
+        if (typeSet.stream().anyMatch(tsDataType -> tsDataType != TSDataType.TEXT)) {
+          throw new SemanticException(
+              "CASE expression: TEXT and other types cannot exist at same time");
+        }
+        return setExpressionType(caseWhenThenExpression, TSDataType.TEXT);
+      }
+      // if BOOLEAN exist, every branch need to be BOOLEAN
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java:323]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java [hunk 15]
```diff
@@ -296,6 +300,52 @@ public TSDataType visitNullOperand(NullOperand nullOperand, Void context) {
...
+      if (typeSet.contains(TSDataType.BOOLEAN)) {
+        if (typeSet.stream().anyMatch(tsDataType -> tsDataType != TSDataType.BOOLEAN)) {
+          throw new SemanticException(
+              "CASE expression: BOOLEAN and other types cannot exist at same time");
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java:327]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java [hunk 16]
```diff
@@ -296,6 +300,52 @@ public TSDataType visitNullOperand(NullOperand nullOperand, Void context) {
...
+      Set<TSDataType> typeSet = new HashSet<>();
+      for (WhenThenExpression whenThenExpression :
+          caseWhenThenExpression.getWhenThenExpressions()) {
+        typeSet.add(process(whenThenExpression, context));
+      }
+      TSDataType elseType = process(caseWhenThenExpression.getElseExpression(), context);
+      if (elseType != null) {
+        typeSet.add(elseType);
+      }
+      // if TEXT exist, every branch need to be TEXT
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java:315]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:61]: The modification has been made as per your suggestion.
UT has been added at org.apache.iotdb.db.mpp.expression.other.CaseWhenThenExpressionTest, is this location appropriate?

File: docs/UserGuide/Operators-Functions/Conditional.md [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+|1970-01-01T08:00:00.001+08:00|         1.0|80.0|
+|1970-01-01T08:00:00.002+08:00|         2.0|70.0|
+|1970-01-01T08:00:00.003+08:00|         3.0|70.0|
+|1970-01-01T08:00:00.004+08:00|         4.0|80.0|
++-----------------------------+------------+----+
+```
+
+##### Example 3
+
+If none of the conditions match and there is no ELSE clause,
```
Comment (lancelly) [inline docs/UserGuide/Operators-Functions/Conditional.md:129]: matches

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+    this.whenThenExpressions = whenThenExpressions;
+    this.elseExpression = elseExpression;
+    if (this.elseExpression == null) {
+      this.elseExpression = new NullOperand();
+    }
+  }
+
+  public CaseWhenThenExpression(ByteBuffer byteBuffer) {
+    int len = ReadWriteIOUtils.readInt(byteBuffer);
+    assert len > 0;
```
Comment (lancelly) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:57]:
```suggestion
   Validate.isTrue(len > 0);
```

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java [hunk 19]
```diff
@@ -296,6 +300,52 @@ public TSDataType visitNullOperand(NullOperand nullOperand, Void context) {
...
+        CaseWhenThenExpression caseWhenThenExpression, Void context) {
+      Set<TSDataType> typeSet = new HashSet<>();
+      for (WhenThenExpression whenThenExpression :
+          caseWhenThenExpression.getWhenThenExpressions()) {
+        typeSet.add(process(whenThenExpression, context));
+      }
+      TSDataType elseType = process(caseWhenThenExpression.getElseExpression(), context);
+      if (elseType != null) {
+        typeSet.add(elseType);
+      }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionTypeAnalyzer.java:314]: done.

File: docs/zh/UserGuide/Operators-Functions/Conditional.md [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+
+## 条件表达式
+
+### CASE
+
+#### 介绍
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply docs/zh/UserGuide/Operators-Functions/Conditional.md:26]: done

File: docs/zh/UserGuide/Operators-Functions/Overview.md [hunk 21]
```diff
@@ -239,4 +239,12 @@ OR, |, ||
...
\ No newline at end of file
+4. 将注册脚本复制到 IoTDB 的程序目录下（与`sbin`目录同级的根目录下），修改脚本中的参数（如果需要）并运行注册脚本以注册 UDF。
+
+## 条件表达式
+
+| 表达式名称                     | 含义        |
+|---------------------------|-----------|
+| `CASE` | 类似if else |
+
+详细说明及示例见文档 [条件表达式](./Conditional.md)
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply docs/zh/UserGuide/Operators-Functions/Overview.md:250]: moved to another place

File: docs/UserGuide/Operators-Functions/Conditional.md [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+    CASE 
+        WHEN caseValue=whenValue1 THEN expression1
+        [WHEN caseValue=whenValue1 THEN expression1] ...
+        [ELSE expression_end]
+    END
+    ```
+
+如果condition均不为真，或均不满足caseVaule=whenValue，则返回expression_end；不存在ELSE子句则返回null。
+
+If none of the conditions are true, or if none of the `whenValue`s match the `caseValue`, the `expression_end` will be returned.
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply docs/UserGuide/Operators-Functions/Conditional.md:64]: done~

File: integration-test/src/test/java/org/apache/iotdb/db/it/query/IoTDBCaseWhenThenIT.java [hunk 23]
```diff
@@ -0,0 +1,426 @@
...
+
+import static org.apache.iotdb.db.it.utils.TestUtils.assertTestFail;
+import static org.apache.iotdb.db.it.utils.TestUtils.prepareData;
+import static org.apache.iotdb.db.it.utils.TestUtils.resultSetEqualTest;
+import static org.apache.iotdb.itbase.constant.TestConstant.DEVICE;
+import static org.apache.iotdb.itbase.constant.TestConstant.TIMESTAMP_STR;
+
+@RunWith(IoTDBTestRunner.class)
+@Category({LocalStandaloneIT.class, ClusterIT.class})
+public class IoTDBCaseWhenThenIT {
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply integration-test/src/test/java/org/apache/iotdb/db/it/query/IoTDBCaseWhenThenIT.java:44]: sure

File: docs/UserGuide/Operators-Functions/Conditional.md (hunk 17 above)
Comment (liyuheng55555) [reply docs/UserGuide/Operators-Functions/Conditional.md:129]: this example has been deleted in new version documentation.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java (hunk 18 above)
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/other/CaseWhenThenExpression.java:57]: indeed.

File: docs/UserGuide/Operators-Functions/Conditional.md [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+  For example, the above SQL statement will be transformed into:
+    ```sql
+    CASE 
+        WHEN caseValue=whenValue1 THEN expression1
+        [WHEN caseValue=whenValue1 THEN expression1] ...
+        [ELSE expression_end]
+    END
+    ```
+
+如果condition均不为真，或均不满足caseVaule=whenValue，则返回expression_end；不存在ELSE子句则返回null。
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply docs/UserGuide/Operators-Functions/Conditional.md:62]: done.

File: docs/UserGuide/Operators-Functions/Conditional.md [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+    ```sql
+    CASE 
+        WHEN condition1 THEN expression1
+        [WHEN condition2 THEN expression2] ...
+        [ELSE expression_end]
+    END
+    ```
+  The `condition`s are evaluated one by one. 
+- The first `condition` that is true will return the corresponding `expression`.
+- 格式2：
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply docs/UserGuide/Operators-Functions/Conditional.md:42]: done.

File: docs/zh/UserGuide/Operators-Functions/Overview.md [hunk 26]
```diff
@@ -231,6 +231,16 @@ OR, |, ||
 | EQUAL_SIZE_BUCKET_OUTLIER_SAMPLE   | INT32 / INT64 / FLOAT / DOUBLE | `proportion`取值范围为`(0, 1]`，默认为`0.1`<br>`type`取值为`avg`或`stendis`或`cos`或`prenextdis`，默认为`avg`<br>`number`取值应大于0，默认`3`| INT32 / INT64 / FLOAT / DOUBLE | 返回符合采样比例和桶内采样个数的等分桶离群值采样                |
 | M4     | INT32 / INT64 / FLOAT / DOUBLE | 包含固定点数的窗口和滑动时间窗口使用不同的属性参数。包含固定点数的窗口使用属性`windowSize`和`slidingStep`。滑动时间窗口使用属性`timeInterval`、`slidingStep`、`displayWindowBegin`和`displayWindowEnd`。更多细节见下文。 | INT32 / INT64 / FLOAT / DOUBLE | 返回每个窗口内的第一个点（`first`）、最后一个点（`last`）、最小值点（`bottom`）、最大值点（`top`）。在一个窗口内的聚合点输出之前，M4会将它们按照时间戳递增排序并且去重。 |
 
+## 其它难以归类的语法
```
Comment (lancelly) [inline docs/zh/UserGuide/Operators-Functions/Overview.md:234]: Perhaps we can delete this hh

Comment (lancelly) [reply docs/zh/UserGuide/Operators-Functions/Overview.md:234]: LGTM~ Congratulations!

**Comment by lancelly** [comment]
> add a new kind of expression

LGTM~ Congratulations!

File: docs/zh/UserGuide/Operators-Functions/Overview.md (hunk 26 above)
Comment (liyuheng55555) [reply docs/zh/UserGuide/Operators-Functions/Overview.md]: deleted.

**Review by lancelly — APPROVED** [review]

**Pull request merged**